src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,282 @@
+  onChange?: (value: T | null) => void;
+  /**
+   * Native input element props.
+   */
+  inputProps?: Omit<React.ComponentPropsWithoutRef<'input'>, 'size'>;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * ComboBox component that allows typing a value to filter the options in dropdown list.
+ * Values can be selected either using mouse clicks or using the Enter key.
+ * @example
+ * <ComboBox
+ *   options={[
+ *     { label: 'Item 1', value: 1 },
+ *     { label: 'Item 2', value: 2 },
+ *     { label: 'Item 3', value: 3 },
+ *   ]}
+ *   onChange={() => {}}
+ * />
+ */
+export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+  const { options, value, onChange, className, inputProps, ...rest } = props;
+
+  // Generate a stateful random id if not specified
+  const [id] = React.useState(() => props.id ?? `iui-${nanoid(10)}`);
+
+  useTheme();
+
+  /** Generates an memoized id for an option, given the index from original list */
+  const getOptionId = React.useCallback(
+    (index: number) =>
+      options[index].id ??
+      `${id}-option${options.findIndex(
+        ({ value }) => value === options[index].value,
+      )}`,
+    [options, id],
+  );
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const menuRef = React.useRef<HTMLUListElement>(null);
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  // Set min-width of menu to be same as input
+  const [minWidth, setMinWidth] = React.useState(0);
+  React.useEffect(() => {
+    if (inputRef.current) {
+      setMinWidth(inputRef.current.offsetWidth);
+    }
+  }, [isOpen]);
+
+  const [filteredOptions, setFilteredOptions] = React.useState(options);
+  const [focusedIndex, setFocusedIndex] = React.useState<number>(-1);
+
+  // Maintain internal selected value state synced with `value` prop
+  const [selectedValue, setSelectedValue] = React.useState<T | undefined>();
+  React.useEffect(() => {
+    setSelectedValue(value);
+  }, [value]);
+
+  // Fire onChange callback every time selected value changes
+  React.useEffect(() => {
+    onChange?.(selectedValue ?? null);
+  }, [selectedValue, onChange]);
+
+  // Controlled input value
+  const [inputValue, setInputValue] = React.useState<string>('');
+  const onInput = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setInputValue(event.target.value);
+      inputProps?.onChange?.(event);
+    },
+    [inputProps],
+  );
+
+  // Update input value when option is selected
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    selectedOption && setInputValue(selectedOption.label);
+  }, [selectedValue, options]);
+
+  // Filter options when input value changes
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    if (
+      inputValue &&
+      inputValue.length > 0 &&
+      isOpen &&
+      selectedOption?.label !== inputValue // don't filter list if option is selected
```

Comment:
I don't understand this line.

Reply:
if an option is selected, we want to show the whole list instead of just that one option. If we only show the one value, then it would be very hard to select a different value

Reply:
Makes sense. Maybe change a comment then a bit.

Reply:
Inverted the if condition and updated comment